### ISSUE TITLE
[Druid] Cat form energy bug - losing full tick of energy on powershift

### ISSUE THREAD

**Issue by trickdaemon** [ValorenWoW/sunstrider-core#4726]
**Describe the bug**

If you shift into cat form from cat form before an energy tick, the subsequent energy tick does not apply as intended until 2 seconds later
**Character name**

Trickdaemon
**To Reproduce**

1. shift into cat form with /cast !cat form (routine power shift macro) 
2.  receive 40 energy from furor talent 
3. no energy received on next energy tick

https://db.endless.gg/?spell=768

**Expected behavior**

Normally you shift into cat form and the energy tick applies even if you were right before receiving energy, leaving you with enough energy to eg shred immediately after a shift.  This way- you can't shred as you have to wait 2 full energy ticks to have enough energy to cast it. 
**Screenshots/videos**

![image](https://user-images.githubusercontent.com/79805241/109507909-3be7c100-7a6d-11eb-9c24-9bfdf59f445e.png)
I shifted at this point in the energy tick, and was not at 60 when the tick completed until 2 seconds later

**_EDIT: Proof:_** 
![image](https://user-images.githubusercontent.com/79805241/109545550-fd66fc00-7a96-11eb-8ee7-06402af63d95.png)

**Additional context**

I don't think this is a latency thing (140ms) as everything else is smooth as butter.  I love the server and I hope this is something that can be fixed, as I was really looking forward to playing cat in pvp and raiding.

Thank you for your time! <3

**Comment by nmagane** [comment]
This is a result of a "fix" that came in patch 2.2.0
The original report is here https://github.com/SunstriderEmu/BugTracker/issues/4665
I wrote a small post about it there just earlier and why I think it shouldn't behave the way it does now.

**Comment by trickdaemon** [comment]
Thank you mmagane! I thought I was going crazy all last night but I can indeed confirm this is not as intended and is something that is wholly detrimental to an already Extremely high effort role.

**Comment by trickdaemon** [comment]
This is a confirmed bug and is not in 2.4.3 patch but was a 2.2 patch behavior that was unintentional by devs.  Energy tick should be clientside and not based on when you enter cat form.   Thank you! sorry for spamming, can't wait to get back into Endless

![image](https://user-images.githubusercontent.com/79805241/109544886-158a4b80-7a96-11eb-8f74-7ed88ae3ce8b.png)

src: https://web.archive.org/web/20080913075822/http://druid.wikispaces.com/ToskksDPSGearMethod#toc6

**Comment by Jerrythemagicdog** [comment]
This footage is post patch 2.3.2 (at least 2.4, as you can see by the Vial of the Sunwell trinket from Magister's Terrace):
https://youtu.be/HVfTsZNMTa8?t=218
https://youtu.be/HVfTsZNMTa8?t=277
https://youtu.be/HVfTsZNMTa8?t=540
https://youtu.be/HVfTsZNMTa8?t=696

There is no mystery here, the entire video can be examined when shifts into Cat Form happen with the same result.

**Comment by trickdaemon** [comment]
Here's the mystery: in the same video you posted, here is an energy tick that occurred before the alleged "proper" 2 second window:
https://youtu.be/HVfTsZNMTa8?t=696
????
Furthermore, worth noting in all of these videos he does not have an energy ticker visible.  To me it seems like a mistake they reverted by 2.4.3. 

Shift has nothing to do with energy regen rate change, that's like saying mana pot affects mana regen rate change.  Those changes are for evocation and adrenaline rush, if it were a bug why do those work and this requires a fix by devs?

Logically makes no sense.

**Comment by nmagane** [comment]
In the clip posted above theres a delay between shifting and getting the 40 energy from furor, this is just a case of spell batching/lag most likely and replicating it doesn't do accomplish anything.

**Comment by trickdaemon** [comment]
There is no energy ticker and none of this can be used as proof without it, to be logical. 

I don't see why this would be patched out or nerfed, it does nothing but harm the enjoyment of the game for people that play cat.


![image](https://user-images.githubusercontent.com/79805241/109605044-88291480-7af2-11eb-9250-f3ebde5e2ad6.png)

**Comment by nmagane** [comment]
It's just a fact that people have played druid with the original ticks for years and years on all TBC pservers, vanilla pservers, and Classic - there is no reason to change it now especially when it's not so clear cut just because of a video with clear serverlag/batching.

**Comment by Evolvee** [comment]
If you knew how spell delay works (and dynamic combat timer in regard to energy tick sync), none of this would suprise you anymore. :shrug_tone5:

**Comment by trickdaemon** [comment]
The paragraph quoted in the alleged bugfix says anything that changes the rate of energy gain which would mean adrenaline rush- furor and shifting does not change the rate of energy gain it just gives you energy just like a mana potion would give you mana.  Does a mana potion reset your five second rule? Oh wait, no it certainly does not.  Match things across the board or fix this, otherwise it's nonsense. 

> Power Regeneration: Any **effect which triggers a change in your rate** of power regeneration (Mana, Rage, Energy, Focus) will now cause an **immediate reward** of some power at the old rate of increase, and then begin new “ticks” of power at the new rate approximately 2 seconds later. This was done to improve functionality of abilities such as Evocation and Innervate so that they did not have wasted “ticks”.
> https://wowwiki.fandom.com/wiki/Patch_2.2.0?oldid=1567509

This is the one of the dumbest things I've experienced in many years, and this behavior will 100% not match TBC classic.

**Comment by Jerrythemagicdog** [comment]
Some more discussion: https://gitlab.com/lights-vengeance/issues/-/issues/368
Some more videos: 
https://youtu.be/Kfy65HJyiac?t=175
https://youtu.be/dc0WrORiHmM?t=90

If it's not working correctly now it shouldn't be hard to prove it, there's plenty of video evidence from TBC.
It's completely understandable why this is annoying, though.

**Comment by nmagane** [comment]
It's not about something being annoying - I would say it's more about expected behavior vs blind replication.
If you look at all the video footage - there is always a delay between shifting and getting the 40 energy from furor in the first place, this does not exist on private servers.
The behavior expected from any druid player that has played most private servers (or TBC before 2.2) does not match what is currently happening,
Nobody is trying to "prove" that that was the way it worked Jerry, but it's more of a broader issue where a very mediocre class is extremely gimped on what is (persumably) the last TBC server before Classic, which also brings up the point that it will most likely not behave like it used to in those videos in Classic TBC either.

I do not think private servers are supposed to be a 1:1 replication of original WoW expansions, and changing something that hundreds of players expected, just for these last few months of pserver TBC, just simply seems like the wrong decision.
Feral will not be "OP" because of this, the PvP ladder will be completely unaffected - This change has had no effect whatsoever in the greater picture - but it's just essentially ruined the game for many of us.
Don't get me wrong when I talk about TBC classic being so soon - I personally would rather stay on Endless rather than play Classic - I just think that this change is taking candy from a baby after letting him have it for over a year on Sunstrider (and pretty much every big server recently?) with no harm resulting of it.

edit:
It is also worth noting that the original reporter of the "bug" (the one spamming  gobal/reacting with laugh emojis here providing 0 input)  stated that it was needed for "PvP reasons" - which frankly couldn't make less sense, considering how there was a total of 1 team playing feral and they were already #1 on a different team.
Ask any top rated Arena player and they will tell you that feral did not need pvp nerfs.

**Comment by trickdaemon** [comment]
100% ^ @nmagane.  The fact that the devs are listening to a troll named penisgodx (guy trolls in game as well) is almost case in point enough that this is a toxic change that just harms player enjoyment of the game.  I thought you didn't support trolls per the discord agreements? This guy has an infantile troll name, and is trolling day by day.  Anyway, I digress: 

Thought experiment; if you want to balance things like this in favor of hard unintentional nerfs of all things, why are paladins afforded both seals? Why are rates 8X? Why give level 30s free mounts to their bags? Etc. - because it's more fun and less punishing on player time/investment/reward. This would be less of a balancing change than double paladin seals is in terms of net effect on balanced gameplay vs other classes and does not affect much of anything outside of making the spec/class more fun.  Feral pvp is considered trash in wow classic even with powershifting being responsive and based on timing.  This is not going to make ferals imba pvp in tbc, just will punish them less for playing the class as intended and make it more responsive with rewarding the timing of casts, not unlike many other classes currently in the game- at an even reduced capacity of power increase. (Re: ice lance with frost bolt on a frozen target for 2 crits, etc.)  As it is, it is a 33% nerf to powershifting making powershifting 1/3 less efficient, not 2x, not 3x, just 1/3 worse and it physically feels awful in terms of timing especially since none of the energy ticking addons can account for the reset. 

I intended to prove this is not how it was as I was under the impression that this wildly detrimental change could not possibly have been live back then, until I discovered this was indeed true and not to mention an unintended change. (Thanks @jerrythemagicdog for those videos)  
I started a big movement in the feral classic discord by digging it up and discovering this and it is going to be 100% petitioned to be changed in TBC classic if it is live pending beta release (doubtful as wow classic is very smooth powershifting.)  I went in one day from 100% hype to play on Endless and about to donate to almost not playing- all over a minute change that you didn't notice until January of this year.  Why not revert to foster player enjoyment in line with all of the other QOL improvements on Endless? Something that is optional and is annoying should not be included in this version of the game, designed to be fun. 


Even a blue post says this was unintentional (as quoted above.) Here's further information from the druid discord (shout out to Mobitz on discord):

> 
> Accidental Powershifting Nerf in TBC 2.4.3
> 
> Apparently, Patch 2.2.0 (September 2007) changed Power Regeneration for abilities such as Evocation and Innervate, which accidentally also led Energy Ticks to always reset when switching into Cat Form, fundamentally changing and nerfing Powershifting in late TBC. While during the majority of TBC, Powershifting behaved as it does in Classic now, it did not in 2.4.3 due to this apparently unintentional change. 
> While this change was not intended, as explained in a Blue Post, it stands to reason that the integral Powershifting Gameplay will be accidentally significantly altered compared to original TBC for the entirety of come Classic TBC, if implemented as expected.
> The Topic was initially dug up by @Trickdaemon.
> Here are the relevant Patch Notes:
> "Power Regeneration: Any effect which triggers a change in your rate of power regeneration (Mana, Rage, Energy, Focus) will now cause an immediate reward of some power at the old rate of increase, and then begin new “ticks” of power at the new rate approximately 2 seconds later. This was done to improve functionality of abilities such as Evocation and Innervate so that they did not have wasted “ticks”."
> https://web.archive.org/web/20070715193106/http://forums.worldofwarcraft.com/thread.html?topicId=174505927&sid=1 (Blue Post Patch Notes)
> https://wowwiki-archive.fandom.com/wiki/Patch_2.2.0 (Wowwiki Patch Notes)
> 
> Blue Post about the same topic, where an equally irritated Moderator also wonders, whether this change is intentional:
> Since the change is on the PTR at the moment and seems to be unrelated to the act of powershifting I would say that the impact on powershifting has been overlooked (either accidentally or intentionally)."
> https://web.archive.org/web/20100528013952/http://forums.wow-europe.com/thread.html?topicId=508511340&sid=1&pageNo=13#252
> 
> Retail Wowhead, Wolfshead Helm: 
> Emblematic for Powershifting, two replies from before and after Patch 2.2.0 on the oldest comment on the Wolfshead Helm sum it up perfectly:
> 
> FroRev on 2007/06/11 (Patch 2.1.1):
> " It stacks with the Furor talent. I always have at least 60 energy upon shifting into Cat Form, and sometimes I have 80 right away if I'm lucky with energy tick timing. It works great with addons like Outfitter. I wear the Wolfshead Helm while I'm in caster form, and Outfitter switches it to an "of the Tiger" helmet as soon as I'm in Cat Form. When I shift out of Cat Form, it switches back to the Wolfshead Helm."
> 
> Twan on 2008/08/01 (Patch 2.4.3):
> "You can't swtich helms during combat and afaik, energy ticks start when you enter Cat Form, so it would take exactly 3 seconds to have 80 energy - it has nothing to do with timing."
> 
> Wowiki change to the Powershifting Article of 2010, referencing that Powershifting was  changed and nerfed in late TBC, while claiming (without source) that this change was intentional.
> https://wowwiki-archive.fandom.com/wiki/Powershifting?diff=prev&oldid=2159813
> 
> Rambling Forum Post from Mid '08, which includes conversations about optimal Rotations. Frequently, people argue whether gaining Energy directly after Powershifting is even possible at all or not, indicating that some were already wise to the implemented changes, while others referenced their own experiences, since the majority of TBC saw the current Classic version of Energy Ticks.
> See Comment (By Lightrain on 2008/06/12 at 9:48 PM (Patch 2.4.1))
> https://www.wowhead.com/forums/topic/26934?page=3
> 
>  Another Forum Post, where User "Mijae" outlines his hypothetical Rotation, which references energy ticks right after Powershifting (the way he remembers it), but others tell him how this has changed. He adapts his hypothetical Rotation accordingly on page 3, which feature Energy Ticks exactly 2 seconds after each Powershifting.
> https://web.archive.org/web/20101109021019/http://elitistjerks.com/f73/t25842-feral_powershifting_macros_thresholds_mana_management/p3/
> 
> Another Forum Post of 2010:
> "Powershiftng for doesn't work like it does with rage anymore. Sometime in BC, i think the very last patch, they changed it so powershifting doesn't give you any more energy than you had before."
> https://www.wowhead.com/forums/topic/140789
> [3:33 PM] Mobitz <Excel Shill>: -
> [3:33 PM] Mobitz <Excel Shill>: WIP Section:
> Information about the 2pT4 Proc resetting the Energy Tick, but partially refunding Energy that was "lost" due to the reset: (This seems to be the intended change of the patch)
> https://web.archive.org/web/20110810071618/http://elitistjerks.com/f31/t11338-feral_druid_dps/p41/
> Rogue perspective:
> https://web.archive.org/web/20070922130058/http://elitistjerks.com/f15/t14287-2_2_ptr_patch_notes_available/p15/

Thank you for your time.

**Comment by ImTheTree** [comment]
I can confirm the issue, on sunstrider it wasnt like this too. even on all the other p-servers.

the energy is ticking with the mana timer, that should not happen. energy ticks should have their own timer.

Please fix so ferals dont feel clunky af, and it kinda kills the dps hard.

**Comment by nmagane** [comment]
Will there be a final answer on this by the dev team? It's been 2 days since the last post here - have you guys discussed this at all or are you still just going by the video? Huge changes like this shouldn't be ignored for too long without closure...

**Comment by trickdaemon** [comment]
Let's try to be patient for them to weigh all of the facts. Though this blue post is all the proof we need to get this reverted if implemented in TBC classic, as well as the applicable note verbiage not remotely applying to furor or shifting into cat form. (Power regeneration rate changes in the game are: spirit tap, innervate, evocate, adrenaline rush, etc. - if furor increased energy regen by 30% for a period of time after shifting this would not be a discussion and would make sense according to the intentional plans of the original TBC dev team.)
![image](https://user-images.githubusercontent.com/79805241/110116010-e53bf900-7d84-11eb-9d4a-e4ec747276db.png)

Appreciate you guys, the server runs very well and I am very impressed despite being bummed about this unintentional powershifting nerf.  Here's a meme from the druid discord and this will be my final say on the matter:  :/
![image](https://user-images.githubusercontent.com/79805241/110116294-4d8ada80-7d85-11eb-9521-505c5f67a846.png)

**Comment by nmagane** [comment]
Again - and not to be annoying and sorry if I am, but please give us a final answer on whether this will be changed back or not, or if the team is still discussing the issue.
 @Jerrythemagicdog

**Comment by Angol36** [comment]
looks like we have to wait for the TBC beta next week and find out there

**Comment by trickdaemon** [comment]
:(

**Comment by kelno** [comment]
I've reviewed this all again, I do believe the current behavior (reset timer) is probably the correct one for 2.4.3 TBC, seeing the various proof + how it works underneath.
But I also agree that this change doesn't look like an intentional nerf by blizzard, this isn't a class that need more nerfing there.  It just looks like a consequence of the change they made about the power regen rate. 

I've made a small exception for cat form and it won't reset the energy timer anymore. The power shifting should work again like you're all used to on pservers.
Applied with next server update.
